Use goreleaser for builds, packaging and release creation

Themimitoof opened the pull request:
This PR changes the way we manage builds, distribution packaging and GitHub releases using [GoReleaser](https://goreleaser.com).

 - Binaries use the same name format as before but make static binaries, strip them and remove debugging bits (see https://github.com/openpubkey/opkssh/issues/99#issuecomment-2776329887 and https://github.com/openpubkey/opkssh/issues/99#issuecomment-2776384599. Will also address #113)
 - Goreleaser creates packages for Debian, RHEL based distributions, Alpine Linux and ArchLinux. 
   - The packaging metadatas can be enhanced if we want to make them available via a package registry like Cloudsmith, Fury or a similar service. 
   - We can also add scripts to bootstrap a server by doing the same steps the `install-linux` does
   - `opkssh` is already packaged by Debian team (https://github.com/openpubkey/opkssh/issues/39 / https://salsa.debian.org/go-team/packages/opkssh). I guess, if we make the package available via a package registry, we could give an upstream release to users not using Debian `sid` (at least for now).

Possible improvements for later:
 - Use cosign: https://www.sigstore.dev
 - Use GoReleaser to publish to `winget` (#95), `Scoop` (#57), ArchLinux `AUR` and `brew`?

Concern #42

EthanHeilman commented:
Thanks for putting this together, I think this will help automate releases and prevent errors. Is there anyway to test the release works as expected without doing a release?

markafarrell commented:
Looks like you can test locally by doing 
`goreleaser release --snapshot --clean`

https://goreleaser.com/quick-start/

Themimitoof commented:
The workflow `build.yml` does exactly this instead of using the `build` command to test if the it creates correctly all artefacts.

EthanHeilman commented:
This is the next PR in line to merge once I test the client side configuration

EthanHeilman commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,29 @@
+          version: "~> v2"
+          args: release --clean --snapshot
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```
What do I need to do to set this up?

Themimitoof replied:
No needs to setup this secret, it is automatically defined when starting a new pipeline: https://docs.github.com/en/actions/security-for-github-actions/security-guides/automatic-token-authentication. The permission of the token is defined with the `permission` stanza.

For this file, the token is not needed as we don't generate a release on GitHub. I will remove it.

EthanHeilman commented:
The main thing this PR is lacking is docs to explain how to use it, what it does, so that everyone could run and maintain the release process.

Themimitoof commented:
I've added documentation for building with GoReleaser to the CONTRIBUTING.md file and revised the release process documentation.

I've also updated the CI/CD pipeline to ensure the CD workflow is only triggered after the CI workflow completes successfully. Unfortunately, as highlighted in the GitHub Actions documentation (https://docs.github.com/en/actions/writing-workflows/choosing-when-your-workflow-runs/events-that-trigger-workflows#workflow_run), I cannot fully test this change until the PR is merged.

EthanHeilman reviewed (approved):
LGTM
